### DISCUSSION TITLE
[Handling Errors] - In production mode, explicit error messages for server components

### DISCUSSION THREAD

**Discussion by Julien-Goletto** [vercel/next.js#62681] [Help] [answered]
### Summary

Context :

NextJS 14.0.3, Typescript, App router
Separated backend API (Django Rest Framework)
Contenerized deployment with docker/docker-compose on a Ubuntu server.
Local dev environment : WSL 2 (Ubuntu), locally deployed contenerized API.

In this application, I wanted to handle errors almost identically, wether they are client or server side.
If client side errors are emitted, I intercept them and play various effects, such as toasting the error message.
As for server components, I use the errors boundaries as per design :

```error.tsx
'use client';
import React from 'react';
import type { Metadata } from 'next';
import { ErrorContent } from '@/app/components/atomic';

export const metadata: Metadata = {
  title: 'My Title',
  description: 'Something went wrong',
};

export default function Error({
  error,
  reset,
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  return <ErrorContent {...{ error, reset }} />;
}
```

As I could'nt extend the Error class to create a custom HttpError class (with status property for instance) and use the extended class with a error.tsx file, I found this kind of work around, aggregating the status in the error message :

```page.tsx
import React from 'react';
import { cookies } from 'next/headers';
import { ssrRequest } from '@/utils';
import { Schema, shipDefaultValuesSchema } from '@/schemas';
import type { Metadata } from 'next';
import { notFound } from 'next/navigation';

export const generateMetadata = async ({
  searchParams,
}: {
  searchParams: { ship_id: string };
}): Promise<Metadata> => {
  try {
    const id = searchParams.id;
    const data= await ssrRequest(
      `${process.env.API_URL}/sata/${id}`,
      schema,
      'GET',
      cookies()
    );
   return data;
    };
  } catch (error) {
    if (error instanceof Error) {
      if (error.message.includes('404')) return notFound();
      throw error;
    }
    throw new Error('Metadata generation failed');
  }
};
```

NextJS censors error message outputs in production as follow :
_An error occurend in the Server Component render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additionnal details about the nature of the error.
Error digest : XXXXXXXXXXXXX_

As I control all kind of emitted errors and some contain essential feedbacks for the user, I wonder if there is a way to perform a clean error handling for server components.

**Comment by tocaherge** [comment]
Losing a bit of hope with the lack of official information or responses on this matter, basically all threads I follow on error tracking are dead ends...

**Reply by Julien-Goletto** [reply]
It's turning me crazy : how such a major topic could have that little coverage ?
Both community / official contents are dodging the bullet.
Are people really shipping NextJS apps in production... ?

**Reply by tocaherge** [reply]
The only way I've found as a "working" setup is having a hacky monkeypatch of the internal logging to forward it to DataDog and then when I get errors or reports of errors I just go have a look in the logs there...
Posted more details in this comment: https://github.com/vercel/next.js/discussions/47563#discussioncomment-9092626

Another thread I've opened on the subject, to no avail: https://github.com/vercel/next.js/discussions/61572

**Comment by SouthLink** [comment]
You can probably use sentry to enhance access to error information by listening for errors in server components and server action
https://docs.sentry.io/platforms/javascript/guides/nextjs/manual-setup/#report-react-component-render-errors

**Reply by tocaherge** [reply]
That's not an option in my projects - and not a reasonable solution for all cases.

**Comment by leerob** [comment]
Can you walk me through a product UI example and the intended output? It sounds like you want to catch an error on the server, and then display that message to the client. Is this for that case, or is it about unhandled exceptions? To confirm, this is _not_ about just viewing the server-side error logs?

**Reply by mertbuilds** [reply]
Hi, thanks for the follow up. My case is very similar to Julien's. 

I have a separate express.js backend which returns me descriptive error messages and I want to show them directly to my users. I'm using server actions to mutate the data but can't show the returned error in my client components (a toast).

For example; I have an onboarding page and if a user tries to register with the same email, my express server handles that very well and returns an error like "This email is already in use".  How can I render this error message in my client component in production?

![CleanShot 2024-06-25 at 00 45 45@2x](https://github.com/vercel/next.js/assets/46185450/753c1fbc-5c85-47e2-9cf7-50fd473227a2)

And if you'd like to try it yourself go to [app.openpurpose.com/new](https://app.openpurpose.com/new) and complete the designer flow twice with a temp mail.

**Reply by leerob** [reply]
Inside if your server action, you can get the response back from your Express API, e.g. "This email is already in use". Then, if the status code is 400 (or whatever logic you want), you can throw `new Error(error.message)" or similar. This will send the message you want to `error.tsx` for display in the client component.

Then, in your client component, you can display the error message by looking at the `error` prop and `error.message`.

Does that work for you?

**Reply by jbrtsnts** [reply]
Hi, thanks for the quick response. We have a similar case: we're using app router and throwing an error using `throw new Error(errorCode)` during data fetching in the `page.tsx` server component.

In local/development, `error.tsx` correctly picks up the `errorCode` in the `error` prop and `error.message`. However, in production, we receive the following message:
> An error occurred in the Server Component render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additional details about the nature of the error. Error digest: XXXXXXXXXXXXX

**Reply by mertbuilds** [reply]
I'm invoking the server action in a button's event handler and I'm using try-catch to catch the error and render a toast. here's the code.

```typescript
try {
  await createDesignerRequest({
    services,
    sampleList,
    email,
  });
} catch (error) {
  if (error instanceof Error) {
    toast(error.message);
  }
}
```

and here is my server action:

```typescript
export async function createDesignerRequest(
  input: Pick<DesignerRequestContextData, "services" | "email" | "sampleList">,
) {
  const body = JSON.stringify({
    services: input.services,
    email: input.email,
    samples: input.sampleList,
  });
  const res = await fetch(`${BACKEND_URL}/designers`, {
    method: "POST",
    body,
    headers: {
      "Content-Type": "application/json",
    },
  });

  const json = await res.json();

  if (!res.ok) {
    throw new Error(json.message);
  }
}
```

I think fundamentally, I'm already doing the same thing you describe but this is resulting with the default error message being rendered in production. it works as expected in development mode.

I'm not sure how using an error.tsx would help me to render a toast with the correct message to be honest. because errors inside error.tsx also being converted to the default error message if they're coming from the server.

![CleanShot 2024-06-25 at 10 17 24@2x](https://github.com/vercel/next.js/assets/46185450/e0d569cb-f0bc-4418-a95d-0d79795a1564)

**Reply by tocaherge** [reply]
To give another use case – when building large internal tooling platforms with a lot of API calls, server actions and other logic on the server in a lot of places I don't necessarily want to handle all errors manually on the server just to format them for an `error.tsx`.
Having error boundaries, both the global root one and specific ones for specific pages/parallel routes, and having control over what they display would help a lot with displaying feedback to the user on errors without needing to clutter up server actions.

If this is a misunderstanding on my part, it seems like the docs could be extended with better examples of how error handling and error tracking of server side logic should be handled.

This also ties into my issues of not being able to find [a reasonable way to pipe logs to Datadog](https://github.com/vercel/next.js/discussions/47563).

**Reply by leerob** [reply]
Thank you! This context is helpful. Taking a look.

**Reply by leerob** [reply]
Okay, so my recommendation here would be to:

1. Inside of your Server Action, you can call your backend API
2. That API returns some error, either manually or caught in a `catch` block
3. The error includes a message you want to show the user
4. Use `useFormState` ([docs](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#server-side-validation-and-error-handling)) (changed to `useActionState` in the [Next.js 15 RC / React 19](https://nextjs.org/blog/next-15-rc)) with your action
5. Return back the error from the server action
6. Display the error client side, or pop it in a toast (whichever UI pattern you prefer)

Does that work for you?

**Reply by mertbuilds** [reply]
So I can't `throw` but I can `return` in server actions if I want to pass my error messages to the client. I can do that without using useFormState if I just do this:

```typescript
export async function createDesignerRequest(
  input: Pick<DesignerRequestContextData, "services" | "email" | "sampleList">,
) {
  const body = JSON.stringify({
    services: input.services,
    email: input.email,
    samples: input.sampleList,
  });
  const res = await fetch(`${BACKEND_URL}/designers`, {
    method: "POST",
    body,
    headers: {
      "Content-Type": "application/json",
    },
  });

  const json = await res.json();

  if (!res.ok) {
    return {
      error: json.message,
    }
  }
}
```
and this,

```typescript

const res = await createDesignerRequest({
  services,
  sampleList,
  email,
});

if (res.error) {
  toast(res.error.message);
}
```

the error mesage shows up in production as well. Your suggestion seems healthier but main point here was using `return` instead of `throw` to carry the error message from server to client. 

Having the `state` from `useFormState` makes it easier to handle the UI so yeah that clarified the situation for me, thanks @leerob!

Since we're using Next.js in production , if I encounter any further issues or any side effects I'll update this thread. 

And to the folks who's reading this, please leave a comment if we're missing something important by using this approach.

**Reply by jbrtsnts** [reply]
Interesting! I can see that using `return` instead of `throw` works in @mert-duzgun's case, but I'm not sure if it would be effective for error boundaries in `error.tsx`.

For instance, we could also consider returning an `errorCode` if `getAccount` fails, like this:

```tsx
export default async function AccountPage({ params }: PageProps) {
    const { data, errorCode } = await getAccount({ accountNumber: params.accountNumber });
    
    if (errorCode) {
        return <Error errorCode={errorCode} />;
    }
                                              
    return <AccountPageContent {...data} />;
}
```

☝️ But this approach seems counterintuitive, and I might be misunderstanding something.

As per the docs if `getAccount` throw an error, which closest `error.tsx` would catch, we just don't want to show a generic `Sorry, something went wrong!` message.

Maybe we could have an option to **NOT** omit the original error message in production builds if we know they are not sensitive information.

**Reply by tocaherge** [reply]
What @jbrtsnts said ☝️ gives more context to my use case mentioned earlier in the thread—having control over what my application displays or not feels valuable.
This would be especially nice in conjunction with parallel routes, for which one of the large selling points for me was the nested error boundary support for displaying errors in a specific part of a complex view.

**Reply by leerob** [reply]
There's two sides here: uncaught exception (error reporting), which should use `error.tsx`, and generalized, caught errors, which should use `useFormState`/ `useActionState`. The latter is how you can take complete control over what the application displays and what information you pass to the client, in response from your server action.

**Reply by Julien-Ose** [reply]
Putting aside server actions particular case, my initial intent was dealing with purposely thrown errors in server components, intercepted by errors boundaries.

Error messages obfuscation in production does not allow us to rely on it for providing consistent feedbacks to our users, unless I missed a point.

**Reply by tocaherge** [reply]
For example to safeguard fetching of something from an API I would need to do something like this now, simplified example:

```tsx
async function getData(): Promise<{ data: SomeData | null, error?: string }> {
  try {
    const apiClient = new ApiClient();
    const data = await apiClient.fetchSomeData({});
    return { data };
  } catch (error) {
    return { data: null, error };
  }
}

export default async function Page() {
  const { data, error } = await getData();

  if (!data) {
    return <ErrorDisplay error={error} />
  }

  return (
    <DataDisplay data={data} />
  );
}
```

When instead it feels like I would much rather have `getData` throw whatever error it wants to, depending on the response—then an `error.tsx` for this page, parallel route, or a parent global boundary would "catch" this error and display the message etc.
This scales a lot better when I have a lot of pages simply fetching data for display, leading to a lot less boilerplate error handling needing to be implemented across multiple files. I could then have different "template" error boundary components for different cases as well.

`page.tsx`
```tsx
async function getData(): Promise<{ data: SomeData | null }> {
  // In a dream world I could trycatch here and re-throw the error with custom data too
  const apiClient = new ApiClient();
  const data = await apiClient.fetchSomeData({});
  return { data };
}

export default async function Page() {
  const { data } = await getData();

  return (
    <DataDisplay data={data} />
  );
}
```

... Then an error boundary handling the display of the eventual error.

Again – I am open to misunderstanding the reasoning behind obfuscation, but in that case I would like to know in a bit more detail *why* I can't have control over the behaviour.

**Reply by RNm-dove** [reply]
I aggree with @tocaherge. If `error.tsx` can handle  error message, we can write a fetch logic more simply.

When `error.tsx` cannot handle error message:
```typescript
export default async function Page() {
  const { data: data1, error: error1 } = await getData1();

  if(!data1) {
    return <ErrorDisplay error={error1} />;
  }

  const { data: data2, error: error2 } = await getData2(data1);

  if(!data2) {
    return <ErrorDisplay error={error2} />;
  }

  const { data: data3, error: error3 } = await getData3(data2);

  if(!data3) {
    return <ErrorDisplay error={error3} />;
  }

  return (
    <DataDisplay data={data3} />
  );
}
```

When `error.tsx` can handle error message:
```typescript
export default async function Page() {
  const { data: data1 } = await getData1();
  const { data: data2 } = await getData2(data1);
  const { data: data3 } = await getData3(data2);

  return (
    <DataDisplay data={data3} />
  );
}
```

**Reply by RNm-dove** [reply]
I came up with a workaround.

I created a wrap function `renderAsyncOrError`.
```typescript
export function renderAsyncOrError(
  rendering: () => Promise<JSX.Element>,
): Promise<JSX.Element> {
  return rendering().catch((error) => {
    if (error instanceof APIError) { // catch your custom error
      return (
        <div>
          <h3>Something went wrong!</h3>
          <div>{error.message}</div>
        </div>
      );
    }

    throw error;
  });
}

```

And, used it every `page.tsx` and `layout.tsx`.

```typescript
export default async function Page() {
  return renderAsyncOrError(() => {
    const { data: data1 } = await getData1();
    const { data: data2 } = await getData2(data1);
    const { data: data3 } = await getData3(data2);
  
    return (
      <DataDisplay data={data3} />
    );
  });
}
```

You can handle custom error.

**Comment by leerob** [comment, marked as answer]
Thank you all for the feedback in this discussion. I've updated the error handling docs to hopefully make this more clear.

https://rc.nextjs.org/docs/app/building-your-application/routing/error-handling

**Reply by Julien-Ose** [reply]
Thanks @leerob for your help.
I was really seduced by the possibilities Error boundaries seemed to offer. Is dealing with crafted errors in a similar way than uncaught ones a part of the upcoming works on Nextjs ?

**Reply by sorenhoyer** [reply]
@leerob I agree about handling known errors without throwing exceptions, but how on earth do you change the status code of your page without throwing an exception? You can't simply show a PDP page with an error and a status code of 200. 

I use neverthrow to get rid of exceptions early in my code, but in the server component, even if I'm not using exceptions, in my switch statement I still have to call `notFound()` to trigger the 404 status code if the error is a `ProductNotFoundError`.

And the problems don't stop there, if I can't use the error.tsx and not-found.tsx. If I wrap an Error boundary around the server component, then whatever is thrown will be caught by that error boundary. But since I can't access the error at all, not even the status code, from inside the boundary component, then I'm unable to show either the NotFound view or the Error view.

So I think the documentation needs to also address this issue, so even for known errors, you should still be able to set the status code.

**Reply by theogravity** [reply]
The problem is the error boundaries / components are all "client"-side rendered, there's no way for us to log unhandled exceptions at all without try / catch + manual log against every possible thing that can be thrown on the server side.

**Reply by 2-fly-4-ai** [reply]
Can't believe this is the state of nextJS in 2026. How do I get the error message? How can I know where to add try/catch when I can't even trace the source of the error. Look at all these unsolved threads.

**Reply by pixelmultiplo** [reply]
Just hit my head on this. What ad odd decision to throw away decades of good practices just to do things differently. I can't wait to move away from next.

**Comment by Julien-Ose** [comment]
Based on @RNm-dove proposal (many thanks), I worked it in a similar way as follow :

```typescript
import React from 'react';
import { ErrorContent } from '@/app/components/atomic';
import { ErrorPageWrapper } from './page-wrapper';
import { notFound } from 'next/navigation';

export const serverComponentErrorBoundary = async (
  fn: () => Promise<JSX.Element>
) => {
  try {
    return await fn();
  } catch (error) {
    if (error instanceof Error) {
      if (error.message.includes('404')) return notFound();
      return <ErrorContent error={error} />;
    }
    throw error;
  }
};
```

I use it as follow :

```typescript
export default async function Page({
  searchParams,
}: {
  searchParams: SearchParams;
}) {
  // Preliminary operations, independent from data fetching

  return serverComponentErrorBoundary(async () => {
    const data1= await getData1();
    const data2= await getData2();
    const data3= await getData3();

    return (
        <PageComponent
          {...{
            seachParams,
            data1,
            data2,
            data3,
          }}
        />
    );
  });
}
```

Each case (weather it is a whole page, a nested page or a component) is quite particular, so the UI chunk fallback has to be adapted for each case.
So the approach is kind of heavy but works perfectly... until NextJS team proposes us something else someday.

Huge thanks to you all.

**Comment by tori-takashi** [comment]
Any update about this shit?

**Comment by satyam-mishra-pce** [comment]
How is there no official way to catch an error on server component and display a message on the client based on the error?
its nuts! especially when its coming from nextjs.

**Reply by 2-fly-4-ai** [reply]
Yes. Unbelievably stupid. Works on vercel so they DGAF.

**Comment by ericmorand** [comment]
I still don't understand how we are supposed to log errors as we want. We are currently stuck with the most basic logging requirement: we need to know what is going wrong in our production system and the only thing that we get from Next.js is:

```sh
[Error: An error occurred in the Server Components render. The specific message is omitted in production builds to avoid leaking sensitive details. A digest property is included on this error instance which may provide additional details about the nature of the error.] {
  digest: '1477216125'
}
```

It seems like the compiler decides how we want to handle errors instead of letting us decide at runtime. What are we missing?

**Comment by deflexable** [comment]
i decided to implement [onRequestError](https://nextjs.org/docs/app/api-reference/file-conventions/instrumentation#onrequesterror-optional) but the error still popup.

Even though onRequestError doesn't run on the client, i genuinely don't know any good reason why next.js is censoring debug data on `onRequestError`.
